### DISCUSSION TITLE
Is there a way to conditionally server side render some content?

### DISCUSSION THREAD

**Discussion by danechitoaie** [vercel/next.js#36542] [Help]
I have a page that I need to render some content if current `date < X` and other content if `date >= X`.
Now if I just to this date check in my page, and have something like:
```javascript
Hello!
{ date > X ? <ComponentA /> : <ComponentB /> }
```
This will "visually work" but both component A and B are still compiled and bundled in the JS when doing next build, right?
So if I have some secret text/string in ComponentB someone looking trough the client side JS files will still find it.

Any idea how to handle this kind of issue?

**Comment by vadymshymko** [comment]
Hello, @danechitoaie. Maybe you should try [next/dynamic](https://nextjs.org/docs/advanced-features/dynamic-import)?

> With it you can import JavaScript modules dynamically and work with them. They also work with SSR.

**Reply by danechitoaie** [reply]
Interesting. I'll try. But `for Next.js to be able to match webpack bundles / module ids to the specific` so des this mean thatmy component will still be publicly accessible as a js bundle? So someone digging trough that could still find my hidden content.

**Reply by vadymshymko** [reply]
I think you are right because a dynamically loaded component is only loaded dynamically and not instantiated

**Comment by icyJoseph** [comment]
Not sure I follow, but you'd have a secret, or sensitive information in the source code? 

While I think this is not very safe, I think if you have the possibility to implement `_middleware`, then you can intercept request coming to `/target-route`, and rewrite them to `/allowed`, `/not-allowed` based on the `date`. 

The nice thing with rewrites is that the browser URL would still be `/target-route`, and in the `_middleware` you can also forbid acess to `/allowed` and `/not-allowed`. Although, come to think of it... there's one way to get around this, and find the component! Hiding things on the client is not safe. Maybe if inside `/not-allowed` was dynamically loaded... but still, it'd be better to fetch the data that hydrates the component from a server, based on the date, and some secret authentication token.

**Reply by danechitoaie** [reply]
I only have a single page, the index page. And I want to organize a contest, a hackathon. The hackathon will start on date X.
So from now until date X I want to sow some general info, rules, signup form. But on date X i want to automatically show the requirements, what the people need to do. So I want this info protected, users should not see that info as if they can "reverse engineer/data mine" it in advance they will start working on the projects before date X.

In PHP for ex, or in a .ejs template rendered on server side I could handle this with a simple if else.

Here in Next.js the only way I can think of is if I have an API route that returns different HTML for that block of code, but doing this, returning HTML from API route and then using `dangerouslySetInnerHTML` feels so dirty and ugly.

Is there a better way?

**Reply by icyJoseph** [reply]
Hi, 

Alright, I was thinking that it was mostly stuff that shouldn't be in the client at all, but for this I'd take this approach.

- Use ISR with `getStaticProps`
   - You could also use `getServerSideProps`, but then you'd render on every request, and without proper caching, this might be less than ideal
- Return `revalidate` set to `1` in `getStaticProps`, perhaps once the event is live, you don't have to revalidate as much, but imagine, if you wanted to release updates to the site, without having to redeploy.

```tsx
// pages/index.tsx
import type { NextPage } from "next";

// Ideally this should come from a CMS/database/S3 bucket, etc...
// but because of `code-elimination` and the way I've written this
// this won't be available client side!
const eventInfo = {
  title: "StartUp Pillow Fight",
  description:
    "Bring your own Pillow, and do something more productive than your 9 to 5.",
  location: "Area 51",
  // any future date
  date: 1651219886097,
};

type EventInfo = typeof eventInfo;

const format = new Intl.DateTimeFormat("en-US");

const EventData = (props: EventInfo) => {
  return (
    <>
      <header>{props.title}</header>

      <main>{props.description}</main>

      <footer>
        {props.location} - {format.format(new Date(props.date))}
      </footer>
    </>
  );
};

type HomeProps = { eventInfo?: EventInfo };

const Home: NextPage<HomeProps> = (props) => {
  if (props.eventInfo) {
    return <EventData {...props.eventInfo} />;
  }

  return (
    <>
      <header>Startup Event!</header>

      <main>More info coming soon, stay tuned!</main>

      <footer>Make sure to come often</footer>
    </>
  );
};

export const getStaticProps = () => {
  const now = new Date();

  const eventTime = new Date(eventInfo.date);

  if (now.getTime() > eventTime.getTime()) {
    // perhaps this could be a longer time, if any at all?
    return { props: { eventInfo }, revalidate: 1 };
  }

  return { props: {}, revalidate: 1 };
};

export default Home;
```

About revalidate:

```md
- Any requests to the page after the initial request and before 1 seconds are also cached and instantaneous.
- After the 1-second window, the next request will still show the cached (stale) page
- Next.js triggers a regeneration of the page in the background.
- Once the page generates successfully, Next.js will invalidate the cache and show the updated page. If the background regeneration fails, the old page would still be unaltered.
```

And the code-elimination result (after stripping the types). [Link to the tool](https://next-code-elimination.vercel.app/).

```js
// This is the code that is bundled for the client-side:
import type { NextPage } from 'next';
const format = new Intl.DateTimeFormat('en-US');
const Event = (props) => {
  return (
    <>
      <header>{props.title}</header>

      <main>{props.description}</main>

      <footer>
        {props.location} - {format.format(new Date(props.date))}
      </footer>
    </>
  );
};
const Home = (props) => {
  if (props.eventInfo) {
    return <Event {...props.eventInfo} />;
  }
  return (
    <>
      <header>Startup Event!</header>

      <main>More info coming soon, stay tunned!</main>

      <footer>Make sure to come often</footer>
    </>
  );
};
export var __N_SSG = true;
export default Home;
```

One down side with `getStaticProps` and revalidate, is that the first user to come after the event release time, still see's the old site. However there's an `unstable_revalidate` feature you could use to manually trigger this yourself, or you could navigate to the site on release date, to make sure everything looks fine.

Another down side is that the EventData component cannot take in a description with nested a tags for example, you'd need to template every single HTML element as JSX.

At least, before my morning coffee this makes sense to me, what do you think? (it is also possible to do something with middlewares and such)

**Reply by danechitoaie** [reply]
Yeah, I'll try this. Thank you.

**Reply by icyJoseph** [reply]
Let us know how it goes. This was a interesting use case.
